### PR TITLE
Improve error messages from bad capture operator usage

### PR DESCRIPTION
**Pull request by josevalim**

### CODE REVIEW

**Review by v0idpwn — APPROVED** [review]

File: lib/elixir/src/elixir_fn.erl [hunk 1]
```diff
@@ -178,18 +178,21 @@ format_error(clauses_with_different_arities) ->
...
+  io_lib:format("block expressions are not allowed inside the capture operator &, got: ~ts",
                 ['Elixir.Macro':to_string(Expr)]);
 format_error({nested_capture, Arg}) ->
-  io_lib:format("nested captures via & are not allowed: ~ts", ['Elixir.Macro':to_string(Arg)]);
+  io_lib:format("nested captures are not allowed. You cannot define a function using "
+    " the capture operator & inside another function defined via &. Got invalid nested "
+    "capture: ~ts", ['Elixir.Macro':to_string(Arg)]);
 format_error({invalid_arity_for_capture, Arity}) ->
-  io_lib:format("invalid arity for &, expected a number between 0 and 255, got: ~b", [Arity]);
+  io_lib:format("capture argument &~B must be a number between 0 and 255", [Arity]);
```
Comment (fertapric) [inline lib/elixir/src/elixir_fn.erl:188]: Minor nitpick:

```suggestion
  io_lib:format("capture argument &~B is invalid (it must be a number between 0 and 255)", [Arity]);
```

Comment (josevalim) [reply lib/elixir/src/elixir_fn.erl:188]: I prefer the current approach as it is more direct. The fact it is an error already reveals it is invalid. WDYT? :)

**Review by whatyouhide — APPROVED** [review]
Yep, all these make things clearer IMO 👍

File: lib/elixir/src/elixir_fn.erl (hunk 1 above)
Comment (fertapric) [reply lib/elixir/src/elixir_fn.erl:188]: What about?

```suggestion
  io_lib:format("capture argument &~B must be numbered between 0 and 255", [Arity]);
```

File: lib/elixir/test/elixir/kernel/expansion_test.exs [hunk 2]
```diff
@@ -1120,7 +1120,7 @@ defmodule Kernel.ExpansionTest do
     end
 
     test "fails on invalid arity" do
-      message = ~r"invalid arity for &, expected a number between 0 and 255, got: 256"
+      message = ~r"capture argument &256 must be a number between 0 and 255"
```
Comment (josevalim) [inline lib/elixir/test/elixir/kernel/expansion_test.exs:1123]:
```suggestion
      message = ~r"capture argument &256 must be numbered between 0 and 255"
```

**Review by fertapric — APPROVED** [review]

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**